src/vmm/src/vstate/vcpu/aarch64.rs

```diff
@@ -67,26 +66,33 @@ impl KvmVcpu {
     ///
     /// * `index` - Represents the 0-based CPU index between [0, max vcpus).
     /// * `vm` - The vm to which this vcpu will get attached.
-    pub fn new(index: u8, vm: &Vm) -> Result<Self, KvmVcpuError> {
-        let kvm_vcpu = vm
-            .fd()
-            .create_vcpu(index.into())
-            .map_err(KvmVcpuError::CreateVcpu)?;
-
-        Ok(KvmVcpu {
+    pub fn new(index: u8, _vm: &Vm) -> Self {
+        KvmVcpu {
             index,
-            fd: kvm_vcpu,
             mmio_bus: None,
             mpidr: 0,
             kvi: None,
-        })
+        }
     }
 
     /// Gets the MPIDR register value.
     pub fn get_mpidr(&self) -> u64 {
         self.mpidr
     }
 
+    /// Runs the vCPU in KVM context and handles the kvm exit reason.
+    ///
+    /// Returns error or enum specifying whether emulation was handled or interrupted.
+    pub fn run_arch_emulation(&self, exit: VcpuExit) -> Result<VcpuEmulation, VcpuError> {
+        METRICS.vcpu.failures.inc();
+        // TODO: Are we sure we want to finish running a vcpu upon
+        // receiving a vm exit that is not necessarily an error?
+        error!("Unexpected exit reason on vcpu run: {:?}", exit);
+        Err(VcpuError::UnhandledKvmExit(format!("{:?}", exit)))
+    }
+}
+
+impl Vcpu {
```

Comment:
I can't say I am in favor of this. With these changes `Vcpu` implementation is spread across `mod.rs` and specific arch files. Same with `KvmVcpu` impl now exists in the `mod.rs`.  So we are basically cooking spaghetti  here at this point.
I did some experimentation and I think simply moving all handling of KVM exits to the arch files will be cleaner.
Example is here (only x86): https://github.com/firecracker-microvm/firecracker/compare/main...ShadowCurse:firecracker:vcpu_exit_process_update?expand=1

This actually makes more sense in my mind as all `KVM` specific implementation will be in the arch specific files, while `mod.rs` and `Vcpu` will only act as a communication wrappers.
The change leaves `fd` in the `KvmVcpu`s, so it leaves these annoying accesses to the fds, but I think we can do a follow up that cleans this with additional methods on `KvmVcpu` side.
The only downside of my approach is that there will be duplicated handling of many `KVMEXIT`s. But at the same time, majority of them have no logic, just set metric and return error, so maybe it is not that bad.

Reply:
I think duplicating the code is strictly worse, because then we'd have the danger of diverging kvm exit handling if we happen to forget to update one of the copies. At this point, I would greatly prefer to just have everything be stand-alone functions (we actually have a ton of those already in arch/aarch64/vcpu.rs, which now makes the wonder where we actually draw the boundary between the vcpu moduel and the arch module?). That said, I think the splitting of the `Vcpu` impls across multiple modules to account for differences due to arch specifics is fine.

For the KvmVcpu impl in `mod.rs` I kinda agree that it's not nice. I think the problem here is that `KvmVcpu` is not actually "kvm-arch specific functionality" like the comment says - mmio_bus is present on both architectures, and not actually used at all in the arch modules! So we could fix this by hoisting `mmio_bus` out of `KvmVcpu`, either via introducing something like 

```rs
struct VcpuEmulationState {
    mmio_bus: Bus,
    arch_vcpu: KvmVcpu
}

impl VcpuEmulationState {
    fn handle_kvm_exit(&mut self, result: Result<...,...>) -> ... { ... }
}
```

inside of `mod.rs`. Or by just moving `mmio_bus` into `Vcpu` and making `handle_kvm_exit` freestanding, e.g. `fn handle_kvm_exit(kvm_vcpu: &mut KvmVcpu, mmio_bus: &mut Bus, result: ...) -> ...`. Happy to do these in this PR, but larger refactors are probably out of scope here, so maybe we can just track those in a follow up?

Reply:
Ok, how about sharing processing function, like this: https://github.com/firecracker-microvm/firecracker/compare/main...ShadowCurse:firecracker:vcpu_exit_process_update
The only nuance is that we will have `#[cfg(target_arch = "x86_64")]` and `#[cfg(target_arch = "aarch64")]` for `IoOut` and `IoIn` exits, but don't think this is a big deal. Technically for `aarch64` these kvm exits should never fire, because `aarch64` does not support pio.
I think the moving of `fd` or `mmio_bus` and other things can be in a separate PR.

Reply:
Okay, I've dropped the VcpuFd moving commit and done something similar to what you proposed (free standing processing function, but still calls down into architecture specific modules)